src/main/kotlin/org/elm/workspace/ElmFormatCLI.kt

```diff
@@ -1,27 +1,82 @@
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.progress.ProgressManager
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.PsiFile
+import org.elm.lang.core.psi.ElmFile
 import org.elm.openapiext.GeneralCommandLine
 import org.elm.openapiext.execute
-import org.elm.openapiext.pathAsPath
-import org.elm.openapiext.withWorkDirectory
+import org.elm.openapiext.isSuccess
+import org.elm.openapiext.toPsiFile
 import java.nio.file.Path
 
 class ElmFormatCLI(val elmFormatExecutablePath: Path) {
 
-    fun reformatFile(project: Project, elmVersion: Version, file: VirtualFile, owner: Disposable = project): ProcessOutput {
+    fun reformatDocument(elmVersion: Version, document: Document): ProcessOutput {
         val arguments = listOf(
                 "--yes",
-                "--elm-version",
-                "${elmVersion.x}.${elmVersion.y}",
-                file.path
+                "--elm-version=${elmVersion.x}.${elmVersion.y}",
+                "--stdin"
         )
+
         return GeneralCommandLine(elmFormatExecutablePath)
-                .withWorkDirectory(file.parent.pathAsPath)
                 .withParameters(arguments)
-                .execute(owner)
+                .execute(document.text)
     }
+
+    companion object {
+
+        fun getElmVersion(project: Project, file: VirtualFile): Version? {
+            val psiFile = file.toPsiFile(project) ?: return null
+            return getElmVersion(psiFile)
+        }
+
+        fun getElmVersion(psiFile: PsiFile): Version? {
+            if (psiFile !is ElmFile) return null
+            val elmVersion: Version = when (val elmProject = psiFile.elmProject) {
+                is ElmApplicationProject -> elmProject.elmVersion
+                is ElmPackageProject -> elmProject.elmVersion.low
+                else -> return null
+            }
+
+            return elmVersion
+        }
+
+        fun formatDocumentAndSetText(ctx: Context) {
+
+            val file = FileDocumentManager.getInstance().getFile(ctx.document)
+            val version = getElmVersion(ctx.project, file!!)
+            val result = ProgressManager.getInstance().runProcessWithProgressSynchronously<ProcessOutput, ExecutionException>({
+                ctx.elmFormatCLI.reformatDocument(version!!, ctx.document)
+            }, "Running elm-format on current file...", true, ctx.project)
+
+            if (result.isSuccess) {
+                val formatted = result.stdout
+                val source = ctx.document.text
+
+                if (source != formatted) {
+                    ApplicationManager.getApplication().runWriteAction {
+                        CommandProcessor.getInstance().runUndoTransparentAction {
+                            ctx.document.setText(formatted)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    data class Context(
```

Comment:
The Context class should be moved back into ElmExternalFormatAction. It’s kind of a UI concern. It exists solely so that the logic for determining whether the action should be enabled or not is the same as the logic for acquiring the data that is necessary to perform the formatting.